keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,346 @@
+* exposing the configured [storage versions][] of Kubernetes built-in resource
+  types and aggregated resources. 
+
+Note that custom resources already exposes this information. 
+
+The migration then is triggered by updates of storage versions.
+
+[Active Management]:https://docs.google.com/document/d/1eoS1K40HLMl4zUyw5pnC05dEF3mzFLp5TPEEt4PFvsM
+[storage versions]:https://github.com/kubernetes/kubernetes/blob/db9545e69e630e5bc8c3c2dab14f9339dd3b871c/pkg/kubeapiserver/options/storage_versions.go#L98
+
+#### b. Using the apiserver version to trigger migration
+
+If the Kubernetes deprecation [policy][] is followed by every apiserver, and
+every cluster uses the default [storage version configs][], then we can use the
+apiserver version to drive the migration. More specifically,
+* kube-apiserver removes the `--storagve-version` flag, so that every deployment
+  of kube-apiserver uses the default storage versions. [#67678][]
+* aggregated apiservers follow semantic versioning, and expose their version via
+  API.
+* aggregated apiservers follow the same API deprecation [policy][].
+
+In both alternatives, the migration of custom resources is triggered by the
+difference between its [storage version][] and the [stored version][].
+
+[storage version configs]:https://github.com/kubernetes/kubernetes/blob/db9545e69e630e5bc8c3c2dab14f9339dd3b871c/pkg/kubeapiserver/options/storage_versions.go#L98
+[storage version]:https://github.com/kubernetes/kubernetes/blob/c1f7df2b0efe89fb9d26db2f2837bac11d57064e/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L66
+[stored version]:https://github.com/kubernetes/kubernetes/blob/c1f7df2b0efe89fb9d26db2f2837bac11d57064e/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L177
+[policy]:https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+[#67678]:https://github.com/kubernetes/kubernetes/pull/67678
+
+In both alternatives, HA apiservers present additional challenges. The migration
+must start after all HA apiservers have agreed on the storage versions or the
+apiserver versions. The agreement protocol is worth its own KEP.
+
+We will revisit this section in 2018 Q4. 
+
+### Risks and Mitigations
+
+The migration process does not change the objects, so it will not pollute
+existing data.
+
+If the rate limiting is not tuned well, the migration can overload the
+apiserver. Users can delete the migration controller and the migration
+jobs to mitigate.
+
+Before upgrading or downgrading the cluster, the cluster administrator MUST
+list the `migration` objects, and make sure every resource type is marked as
+completed. Otherwise the apiserver can crash. To mitigate, the cluster
+administrator can rollback the apiserver to the old version, and wait for the
+migration to complete. If the apiserver does not crash, the `migration` will be
+corrupted, administrator needs to re-launch the **kube-migrator** to recover.
```

Comment:
We are concerned that there is no one-to-one mapping between apiserver version and storage versions, because user can configure `--storage-versions` flag, though the flag will be removed in 1.13.
Another concern is that the aggregated apiserver doesn't have a server version. If we can agree on either forcing aggregated apiserver follows semver, or that the storage migration doesn't handle aggregated APIs by default, we can implement what you suggested.

Reply:
> it is not clear why 'migration' will be corrupted

Expanded. "If the apiserver does not crash after upgrading or downgrading, the `migration` objects lose track of migration progress, because the default storage versions might changed after upgrading".
or downgrading, but no one increments the `migration.spec.generation`.